plugins/blockvolume/blockprovider/gluster-block/volume.go

```diff
@@ -0,0 +1,163 @@
+package glusterblock
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/rand"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/gluster/glusterd2/glusterd2/commands/volumes"
+	"github.com/gluster/glusterd2/glusterd2/gdctx"
+	"github.com/gluster/glusterd2/glusterd2/transaction"
+	"github.com/gluster/glusterd2/glusterd2/volume"
+	"github.com/gluster/glusterd2/pkg/api"
+	"github.com/gluster/glusterd2/pkg/size"
+
+	"github.com/pborman/uuid"
+)
+
+// BlockSizeFilter returns a volume Filter, which will filter out volumes
+// haing block-hosting-available-size greater than give size.
+func BlockSizeFilter(size int64) volume.Filter {
+	return func(volinfos []*volume.Volinfo) []*volume.Volinfo {
+		var volumes []*volume.Volinfo
+
+		for _, volinfo := range volinfos {
+			availableSize, found := volinfo.Metadata["_block-hosting-available-size"]
+			if !found {
+				continue
+			}
+
+			if availableSizeInBytes, err := strconv.Atoi(availableSize); err == nil && int64(availableSizeInBytes) > size {
+				volumes = append(volumes, volinfo)
+			}
+		}
+		return volumes
+	}
+}
+
+// GetExistingBlockHostingVolume returns a existing volume which is suitable for hosting a gluster-block
+func GetExistingBlockHostingVolume(size int64) (*volume.Volinfo, error) {
+	var (
+		filters     = []volume.Filter{volume.FilterBlockHostedVolumes, BlockSizeFilter(size)}
+		ctx, cancel = context.WithTimeout(context.Background(), time.Second*5)
+	)
+
+	defer cancel()
+
+	volumes, err := volume.GetVolumes(ctx)
+	if err != nil || len(volumes) == 0 {
+		return nil, fmt.Errorf("%v/no volumes found", err)
+	}
+
+	volumes = volume.ApplyCustomFilters(volumes, filters...)
+
+	return SelectRandomVolume(volumes)
+}
+
+// CreateBlockHostingVolume will create a gluster volume with metadata block-hosting-volume-auto-created=yes
+func CreateBlockHostingVolume(req *api.VolCreateReq) (*volume.Volinfo, error) {
+	status, err := volumecommands.CreateVolume(context.Background(), *req)
+	if err != nil || status != http.StatusCreated {
+		return nil, err
+	}
+	vInfo, err := volume.GetVolume(req.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	vInfo.Metadata["_block-hosting-volume-auto-created"] = "yes"
```

Comment:
Using `_` prefix will make it impossible to manually create block hosting volume and use that for block file creation.(Since Metadata keys starts with `_` are restricted for API users). We can skip using prefix or we can expose BlockHosting=Yes/No flag in Volume Edit API